File: 2022/Day4.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def parse(inp):
    return [(set(range(a, b+1)), set(range(c,d+1))) for a,b,c,d in map(pnums, lines(inp))]

def p1(inp):
    return sum(1 for a,b in inp if a.issubset(b) or b.issubset(a))

def p2(inp):
    return sum(1 for a,b in inp if len(a.intersection(b)) != 0)

def main():
    inp = parse(data())
    print(f"Solution for part 1:\n{p1(inp)}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 450
# Solution part 2: 837

File: 2022/Day4.hs
#! /usr/bin/env -S runhaskell
import Data.List
import Data.List.Split
import Data.Char

import Util

type Input = [([Int], [Int])]
type Output = Int

parse :: String -> Input
parse = map (tuple2 . map (uncurry enumFromTo . tuple2 . map read . splitOn "-") . splitOn ",") . lines

part1 :: Input -> Output
part1 = count $ or . (uncurry <$> ([id, flip] <*> pure issubset) <*>) . pure

part2 :: Input -> Output
part2 = count $ not . null . uncurry intersect

main :: IO()
main = do
    f <- readFile("4.in")
    let input = parse f
    putStr "Solution for part 1: "
    putStrLn . show $ part1 input
    putStr "Solution for part 2: "
    putStrLn . show $ part2 input

-- Solution part 1: 450
-- Solution part 2: 837
